text/0000-ptr-meta.md

```diff
@@ -0,0 +1,387 @@
+/// The data pointer can be extracted by casting a (fat) pointer
+/// to a (thin) pointer to a `Sized` type the `as` operator,
+/// for example `(x: &dyn SomeTrait) as *const SomeTrait as *const ()`.
+///
+/// [dst]: https://doc.rust-lang.org/nomicon/exotic-sizes.html#dynamically-sized-types-dsts
+#[lang = "pointee"]
+pub trait Pointee {
+    /// The type for metadata in pointers and references to `Self`.
+    type Metadata: Copy + Send + Sync + Ord + Hash + Unpin + 'static;
+}
+
+/// Pointers to types implementing this trait alias are “thin”:
+///
+/// ```rust
+/// fn this_never_panics<T: std::ptr::Thin>() {
+///     assert_eq!(std::mem::size_of::<&T>(), std::mem::size_of::<usize>())
+/// }
+/// ```
+pub trait Thin = Pointee<Metadata=()>;
+
+/// Extract the metadata component of a pointer.
+///
+/// Values of type `*mut T`, `&T`, or `&mut T` can be passed directly to this function
+/// as they implicitly coerce to `*const T`.
+/// For example:
+///
+/// ```
+/// assert_eq(std::ptr::metadata("foo"), 3_usize);
+/// ```
+///
+/// Note that the data component of a (fat) pointer can be extracted by casting
+/// to a (thin) pointer to any `Sized` type:
+///
+/// ```
+/// # trait SomeTrait {}
+/// # fn example(something: &SomeTrait) {
+/// let object: &SomeTrait = something;
+/// let data_ptr = object as *const SomeTrait as *const ();
+/// # }
+/// ```
+pub fn metadata<T: ?Sized>(ptr: *const T) -> <T as Pointee>::Metadata {…}
+
+impl<T: ?Sized> *const T {
+    pub fn from_raw_parts(data: *const (), meta: <T as Pointee>::Metadata) -> Self {…}
+}
+
+impl<T: ?Sized> *mut T {
+    pub fn from_raw_parts(data: *mut (), meta: <T as Pointee>::Metadata) -> Self {…}
+}
+
+impl<T: ?Sized> NonNull<T> {
+    pub fn from_raw_parts(data: NonNull<()>, meta: <T as Pointee>::Metadata) -> Self {
+        unsafe {
+            NonNull::new_unchecked(<*mut _>::from_raw_parts(data.as_ptr(), meta))
+        }
+    }
+}
+
+/// The metadata for a `dyn SomeTrait` trait object type.
+///
+/// It is a pointer to a vtable (virtual call table)
+/// that represents all the necessary information
+/// to manipulate the concrete type stored inside a trait object.
+/// The vtable notably it contains:
+///
+/// * type size
+/// * type alignment
+/// * a pointer to the type’s `drop_in_place` impl (may be a no-op for plain-old-data)
+/// * pointers to all the methods for the type’s implementation of the trait
+///
+/// Note that the first three are special because they’re necessary to allocate, drop,
+/// and deallocate any trait object.
+///
+/// The layout of vtables is still unspecified, so this type is a more-type-safe
+/// convenience for accessing those 3 special values. Note however that `DynMetadata` does
+/// not actually know the trait it’s associated with, indicating that, at very least,
+/// the location of `size`, `align`, and `drop_in_place` is identical for all
+/// trait object vtables in a single program.
+#[derive(Copy, Clone)]
+pub struct DynMetadata {
```

Comment:
One potential improvement I see with `DynMetadata` is that it does two levels of erasure: 1. the erasure from `dyn` and 2. the erasure from a specific `dyn Trait` to `dyn *`.

If we would keep the specific `dyn Trait` we might in the future add trait specific vtable based methods.

(This would mean `struct DynMetadata<T:?Sized>` e.g. `DynMetadata<dyn Debug>` )

But I'm not sure if this is worth the effort and added complexity.

Reply:
This was discussed before (when DynMetadata still was VTable.

One of the major arguments against it was the potential increased compiler time. (Me guessing: Which given that this is a auto trait might be non neglibale?? Maybe?)

I wonder if we have a way to erase the exact type of DynMetadata having just `impl DynMetadata` instead. But I guess this would require (existential) `impl Trait` features we currently don't have to be doable and nicely usable.

Reply:
I’ve added an unresolved question.